File: solutions/go/125.go
func isPalindrome(s string) bool {
	for i, j := 0, len(s)-1; i < j; /* empty */ {
		if isAlnum(s[i]) && isAlnum(s[j]) {
			if toLower(s[i]) == toLower(s[j]) {
				i++
				j--
			} else {
				return false
			}
		} else if !isAlnum(s[i]) {
			i++
		} else if !isAlnum(s[j]) {
			j--
		} else {
			i++
			j--
		}
	}

	return true
}

func isAlnum(b byte) bool {
	if (b >= '0' && b <= '9') || (b >= 'A' && b <= 'Z') || (b >= 'a' && b <= 'z') {
		return true
	}

	return false
}

func toLower(b byte) byte {
	if b >= 'A' && b <= 'Z' {
		return b + 'a' - 'A'
	}

	return b
}

File: solutions/c/125.c
#include <ctype.h>

bool
isPalindrome(char *s)
{
    char *tail = s;

    while (*tail) {
        ++tail;
    }

    --tail;

    while (s < tail) {
        if (isalnum(*s) && isalnum(*tail)) {
            if (tolower(*s) == tolower(*tail)) {
                ++s, --tail;
            } else {
                return false;
            }
        } else if (!isalnum(*s)) {
            ++s;
        } else if (!isalnum(*tail)) {
            --tail;
        } else {
            ++s, --tail;
        }
    }

    return true;
}
